### PR TITLE
[1/n] chore: decouple quantization implementation from vLLM dependency

### PR DESCRIPTION
**Pull request by AniZpZ**
## Motivation
The primary goal of this change is to enhance the consistency and stability of SGLang's quantization features. By decoupling the quantization implementation from its vLLM dependency, we aim to make the module easier to maintain and more portable.
Full realization of this goal will involve several subsequent PRs; this particular PR addresses the GPTQ feature.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by ch-wan** [comment]
@AniZpZ Awesome! This is exactly what I need to pass the vLLM dependency test in #7966. Could you please share your plan and timeline? Also, let me know it there's something that I can help with.

**Comment by zhyncs** [comment]
Hi @ch-wan JFYI @AniZpZ is my former colleague. He will take over all aspects of quantization in SGLang, including removing dependencies and cleaning up the code. We can sync on Slack.

File: python/sglang/srt/layers/quantization/gptq.py [hunk 1]
```diff
@@ -1,47 +1,52 @@
 import logging
+from dataclasses import dataclass
 from fractions import Fraction
 from typing import Any, Callable, Dict, List, Optional, Union
 
 import torch
-
-from sglang.srt.layers.linear import LinearBase, set_weight_attrs
+from sgl_kernel import fused_marlin_moe
```
Comment (zhyncs) [inline python/sglang/srt/layers/quantization/gptq.py]: We should only import `sgl_kernel` when `_is_cuda` is true

File: python/sglang/srt/layers/quantization/marlin_utils.py [hunk 2]
```diff
@@ -0,0 +1,781 @@
+# SPDX-License-Identifier: Apache-2.0
+# Adapted from https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/utils/marlin_utils.py
+
+import logging
+from typing import Any, Optional
+
+import numpy
+import torch
+from sgl_kernel.scalar_type import ScalarType, scalar_types
```
Comment (zhyncs) [inline python/sglang/srt/layers/quantization/marlin_utils.py]: same as above

File: python/sglang/srt/layers/quantization/utils.py [hunk 3]
```diff
@@ -1,9 +1,11 @@
 # Adapted from https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/layers/quantization/utils/quant_utils.py
 
 from types import MappingProxyType
-from typing import List, Mapping, Tuple, Union
+from typing import List, Mapping, Optional, Tuple, Union
 
+import numpy
 import torch
+from sgl_kernel.scalar_type import ScalarType
```
Comment (zhyncs) [inline python/sglang/srt/layers/quantization/utils.py]: same

**Review by zhyncs — CHANGES_REQUESTED** [review]

**Comment by zhyncs** [comment]
BTW please fix https://github.com/sgl-project/sglang/actions/runs/16248385910/job/45888665139?pr=7992
Thanks for this work!

File: python/sglang/srt/layers/quantization/utils.py (hunk 3 above)
Comment (AniZpZ) [reply python/sglang/srt/layers/quantization/utils.py]: done

**Comment by AniZpZ** [comment]
> BTW please fix https://github.com/sgl-project/sglang/actions/runs/16248385910/job/45888665139?pr=7992 Thanks for this work!

there is a problem that Qwen1.5-1.8B-Chat-GPTQ-4bits-dynamic-cfg-with-lm_head-symFalse is a bf16 model while currently non-marlin gptq kernel only support fp16, the cast may lead to overflow

**Comment by AniZpZ** [comment]
@ch-wan @zhyncs  the vllm dependency test problem has been fixed, could you please further review the PR?

File: test/srt/test_gptqmodel_dynamic.py
Comment (ch-wan) [inline test/srt/test_gptqmodel_dynamic.py:1]: Should we move this test from vllm dependency test to a regular test?

File: python/sglang/srt/layers/quantization/__init__.py
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/__init__.py:1]: Could you confirm if we should remove line 329/330/333 from this file?

File: python/sglang/srt/layers/quantization/compressed_tensors/__init__.py
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/compressed_tensors/__init__.py:1]: I guess it should be deferred to future PRs.

File: python/sglang/srt/layers/quantization/gptq.py [hunk 4]
```diff
@@ -1,48 +1,56 @@
...
+    marlin_repeat_scales_on_all_ranks,
+    marlin_sort_g_idx,
+    marlin_zero_points,
+    verify_marlin_supported,
+)
+from sglang.srt.layers.quantization.scalar_type import ScalarType, scalar_types
+from sglang.srt.layers.quantization.utils import replace_parameter, unpack_cols
 
 try:
     from vllm import _custom_ops as ops
```
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/gptq.py:40]: QPTQLinear still needs vllm. Is it a mistake or we will support it in the future PR?

File: python/sglang/srt/layers/quantization/gptq.py [hunk 5]
```diff
@@ -151,7 +191,7 @@ def from_config(cls, config: Dict[str, Any]) -> "GPTQConfig":
 
     def get_quant_method(
         self, layer: torch.nn.Module, prefix: str
-    ) -> Optional[GPTQLinearMethod]:
+    ) -> Optional["QuantizeMethodBase"]:
```
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/gptq.py]: how can this func return a quant method other than GPTQLinearMethod?

File: python/sglang/srt/layers/quantization/marlin_utils.py [hunk 6]
```diff
@@ -0,0 +1,781 @@
...
+    PackedvLLMParameter,
+)
+from sglang.srt.layers.quantization.base_config import QuantizationConfig
+from sglang.srt.layers.quantization.scalar_type import ScalarType, scalar_types
+from sglang.srt.layers.quantization.utils import pack_cols, unpack_cols
+from sglang.srt.layers.vocab_parallel_embedding import ParallelLMHead
+from sglang.srt.utils import get_device_capability
+
+try:
+    from vllm import _custom_ops as ops
```
Comment (ch-wan) [inline python/sglang/srt/layers/quantization/marlin_utils.py:24]: Again, it still depends on vllm.

File: test/srt/test_int4_kernel.py
Comment (ch-wan) [inline test/srt/test_int4_kernel.py:1]: why is it removed?

File: test/srt/test_w4a8.py
Comment (ch-wan) [inline test/srt/test_w4a8.py:1]: why is it removed?

Comment (AniZpZ) [reply test/srt/test_w4a8.py:1]: This test was introduced by https://github.com/sgl-project/sglang/pull/7683, and was planned to be a unit test for W4AInt8 quantization. However, https://github.com/sgl-project/sglang/pull/7772 introduced W4AFP8 quantization, and we decided to hold the W4AInt8 plan.

File: test/srt/test_int4_kernel.py
Comment (AniZpZ) [reply test/srt/test_int4_kernel.py:1]: same reason as https://github.com/sgl-project/sglang/pull/7992#discussion_r2206180332

File: python/sglang/srt/layers/quantization/gptq.py (hunk 4 above)
Comment (AniZpZ) [reply python/sglang/srt/layers/quantization/gptq.py:40]: Currently, only the non-Marlin GPTQ method depends on vLLM. This is a rare use case, as most models can now apply the faster Marlin method.
This PR focuses on the Marlin method. I will add another PR to introduce a non-Marlin GPTQ kernel and completely remove the dependency.

File: test/srt/test_gptqmodel_dynamic.py
Comment (AniZpZ) [reply test/srt/test_gptqmodel_dynamic.py:1]: That's a good suggestion. As soon as we introduce the non-Marlin GPTQ kernel in SGLang, we can move this test to regular test suite.
I have a branch working on this https://github.com/AniZpZ/sglang/tree/add_gptq_marlin_kernel

File: python/sglang/srt/layers/quantization/marlin_utils.py (hunk 6 above)
Comment (AniZpZ) [reply python/sglang/srt/layers/quantization/marlin_utils.py:24]: same reason as https://github.com/sgl-project/sglang/pull/7992#discussion_r2206185020

File: python/sglang/srt/layers/quantization/gptq.py (hunk 5 above)
Comment (AniZpZ) [reply python/sglang/srt/layers/quantization/gptq.py]: might be UnquantizedLinearMethod

Comment (ch-wan) [reply python/sglang/srt/layers/quantization/gptq.py]: Let's use `LinearMethodBase` or `Union[GPTQLinearMethod, UnquantizedLinearMethod]`

File: python/sglang/srt/layers/quantization/compressed_tensors/__init__.py
Comment (AniZpZ) [reply python/sglang/srt/layers/quantization/compressed_tensors/__init__.py:1]: fixed

File: python/sglang/srt/layers/quantization/__init__.py
Comment (AniZpZ) [reply python/sglang/srt/layers/quantization/__init__.py:1]: Not for now, I plan to remove the patch func when we fully decouple quantization impelementation of compressed tensor and awq

File: python/sglang/srt/layers/quantization/gptq.py (hunk 5 above)
Comment (AniZpZ) [reply python/sglang/srt/layers/quantization/gptq.py]: fixed
